cmd/testpeers.go

```diff
@@ -258,46 +460,132 @@ func runSelfTest(ctx context.Context, queuedTestCases []testCaseName, allTestCas
 	}
 }
 
-func peerPingTest(ctx context.Context, _ *testPeersConfig, _ string) testResult {
-	// TODO(kalo): implement real ping
-	select {
-	case <-ctx.Done():
-		return testResult{Verdict: testVerdictFail}
-	default:
-		return testResult{
-			Verdict: testVerdictFail,
-			Error:   errors.New("ping not implemented").Error(),
+func peerPingTest(ctx context.Context, _ *testPeersConfig, tcpNode host.Host, peer p2p.Peer) testResult {
+	tr := testResult{Name: "Ping"}
+
+	ticker := time.NewTicker(3 * time.Second)
+	defer ticker.Stop()
+
+	for ; true; <-ticker.C {
+		select {
+		case <-ctx.Done():
+			tr.Verdict = testVerdictFail
+			tr.Error = errTimeoutInterrupted.Error()
+
+			return tr
+		default:
+			result, err := pingPeerOnce(ctx, tcpNode, peer)
+			if err != nil {
+				tr.Verdict = testVerdictFail
+				tr.Error = err.Error()
```

Comment:
Looks like you always put Error's text as is to `tr.Error`. Because of this I suspect `tr.Error` shall be of type Error. Final conversion to string can be done in the output renderer (console | toml).